File: LeetCode/297. 二叉树的序列化与反序列化.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    SEP = ','
    NULL = '#'
    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        result = []
        def traverse(node):
            if not node:
                result.append(self.NULL)
                result.append(self.SEP)
                return
            result.append(str(node.val))
            result.append(self.SEP)
            traverse(node.left)
            traverse(node.right)
        traverse(root)
        return ''.join(result)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        nodes = data.split(self.SEP)
        def traverse(nodes):
            if len(nodes) == 0:
                return None
            nodeVal = nodes.pop(0)
            if nodeVal == self.NULL:
                return None
            left = traverse(nodes)
            right = traverse(nodes)
            return TreeNode(nodeVal, left, right)
        return traverse(nodes)
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: LeetCode/297. 二叉树的序列化与反序列化.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
const SEP = ',';
const NULL = '#';
/*
 * Encodes a tree to a single string.
 */
function serialize(root: TreeNode | null): string {
    let result = [];
    const traverse = (node: TreeNode | null) => {
        if (node === null) {
            result.push(NULL,SEP);
            return;
        }
        result.push(node.val, SEP);
        traverse(node.left);
        traverse(node.right);
    }
    traverse(root);
    return result.join('');
};

/*
 * Decodes your encoded data to tree.
 */
function deserialize(data: string): TreeNode | null {
    const nodes = data.split(',');
    const traverse = (nodes: string[]): TreeNode | null => {
        if (nodes.length === 0) return null;
        const rootVal = nodes.shift();
        if (rootVal === NULL) return null;
        const root = new TreeNode(Number(rootVal));
        const left = traverse(nodes);
        const right = traverse(nodes);
        root.left = left;
        root.right = right;
        return root;
    }
    return traverse(nodes);
};


/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */
